File: Leetcode/python/sol797.py

# https://leetcode.com/problems/all-paths-from-source-to-target/description/

class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        end_goal = len(graph) - 1
        S = [(0,[0])]
        ans = []
        while S:
            node, path = S.pop()
            if node == end_goal:
                ans.append(path)
            for j in graph[node]:
                S.append((j, path+[j]))
        return ans

File: Leetcode/cpp/sol797.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int ending = graph.size()-1; // find the ending node
        vector<pair<int, vector<int>>> S; // vector holding the pairs (current node, path so far)
        S.push_back({0, {0}});

        vector<vector<int>> result;

        while(!S.empty()){ // using
            auto temp = S.back(); // gets pair (int, path)
            S.pop_back();
            if(ending == temp.first){ // if ending found, add path and skip a loop (no need to look for the end node's children)
                result.push_back(temp.second);
                continue;
            }

            for(auto i:graph[temp.first]){ // add each children from the current node to the stack
                auto new_node = temp.second;
                new_node.push_back(i);
                S.push_back({i, new_node});
            }
        }

        return result;
    }
};